### PR TITLE
feat/ICL-109: allow to use private key to login in browser env

### PR DESCRIPTION
**Pull request by dawidgil**
This change allows to use private key to login through frontend application. Current implementation allows to use private key to login only in Node env. This is needed to start e2e testing with private key, to omit metamask login.

### CODE REVIEW

File: src/iam/iam-base.ts [hunk 1]
```diff
@@ -180,7 +180,8 @@ export class IAMBase {
         walletProvider?: WalletProvider;
     }): Promise<AccountInfo | undefined> {
         const { privateKey, rpcUrl } = this._connectionOptions;
-        if (this._executionEnvironment === ExecutionEnvironment.NODE) {
+        const isPrivateKeyInBrowser = this._executionEnvironment === ExecutionEnvironment.BROWSER && privateKey;
```
Comment (JGiter) [inline src/iam/iam-base.ts:183]: `isPrivateKeyInBrowser` and `!privateKey` contradicts each other
```suggestion
       if (privateKey) { ... signer from private key}
      if (this._executionEnvironment == ExecutionEnvironment.BROWSER) { throw new Error(ERROR_MESSAGES.PRIVATE_KEY_NOT_PROVIDED)
```

File: src/iam/iam-base.ts [hunk 2]
```diff
@@ -298,7 +299,10 @@ export class IAMBase {
      * @returns boolean that indicates the session state
      */
     public isSessionActive() {
-        return Boolean(this._publicKey) && Boolean(this._providerType);
+        return (
+            (Boolean(this._publicKey) && Boolean(this._providerType)) ||
```
Comment (JGiter) [inline src/iam/iam-base.ts:303]: Is this change because of missing provider of type Wallet? If so then it would be more consistent to add this type

**Review by JGiter — CHANGES_REQUESTED** [review]

File: src/iam/iam-base.ts [hunk 3]
```diff
@@ -1,18 +1,18 @@
-import { providers, Signer, utils, Wallet, BigNumber, ethers } from "ethers";
+import { BigNumber, ethers, providers, Signer, utils, Wallet } from "ethers";
```
[the remark this answers is not in the document]
Comment (dawidgil) [reply src/iam/iam-base.ts:1]: No, i will revert that.

File: src/iam/iam-base.ts (hunk 2 above)
Comment (dawidgil) [reply src/iam/iam-base.ts:303]: There is no wallet type which means there is no providerType. What should i put there? Because localStorage is used only by browser and i can't use only publicKey.

File: src/iam/iam-base.ts (hunk 1 above)
Comment (dawidgil) [reply src/iam/iam-base.ts:183]: Ok, so you prefer to create below an if statement that uses the same part of code?

Comment (JGiter) [reply src/iam/iam-base.ts]: I mean that if private key is provided, then it should be used to create signer regardless of any other conditions

File: src/iam/iam-base.ts (hunk 2 above)
Comment (JGiter) [reply src/iam/iam-base.ts:312]: You can add Wallet providerType. And to avoid storing session in Node.js environment you can check environment before `storeSession`

File: src/iam/iam-base.ts [hunk 4]
```diff
@@ -181,14 +181,13 @@ export class IAMBase {
     }): Promise<AccountInfo | undefined> {
         const { privateKey, rpcUrl } = this._connectionOptions;
         if (this._executionEnvironment === ExecutionEnvironment.NODE) {
```
Comment (JGiter) [inline src/iam/iam-base.ts:183]: I think that from now on we don't need to consider execution environment to initialize signer, so you can remove this block

File: src/iam/iam-base.ts [hunk 5]
```diff
@@ -181,14 +181,13 @@ export class IAMBase {
...
-            if (!rpcUrl) {
-                throw new Error(ERROR_MESSAGES.RPC_URL_NOT_PROVIDED);
-            }
-            this._provider = new JsonRpcProvider({ url: rpcUrl });
-            this._signer = new Wallet(privateKey, this._provider);
+            this.initWithPrivateKey(privateKey, rpcUrl);
+            return;
+        }
+
+        if (WalletProvider.PrivateKey === walletProvider) {
```
Comment (JGiter) [inline src/iam/iam-base.ts:188]: For consistency
```suggestion
        if (walletProvider === WalletProvider.PrivateKey) {
```

**Review by JGiter — CHANGES_REQUESTED** [review]

File: src/iam/iam-base.ts (hunk 4 above)
Comment (dawidgil) [reply src/iam/iam-base.ts:183]: But this will make a breaking change, because Node user will need to set WalletProvider while initializingConnection, am I right?
What's more, for `WalletProvider.PrivateKey` i'm setting `this._providerType = walletProvider;`. Can it also be set for Node env?

Comment (JGiter) [reply src/iam/iam-base.ts:183]: For now we can set providerType can be set in constructor if node environment is detected

**Comment by JGiter** [comment]
Please stick with conventional-commits spec

File: src/iam/iam-base.ts [hunk 6]
```diff
@@ -340,7 +357,7 @@ export class IAMBase {
                 break;
             }
             default:
-                throw new Error("Event not supported");
+            // throw new Error("Event not supported");
```
Comment (JGiter) [inline src/iam/iam-base.ts:360]: Why this change?

File: test/initializeConnection.testSuite.ts [hunk 7]
```diff
@@ -6,10 +6,6 @@ import { WalletProvider } from "../src/types/WalletProvider";
 const iam_withoutKey = new IAM({ rpcUrl });
 
 export const initializeConnectionTests = () => {
-    test("initializeConnection requires privateKey or walletProvider enum", async () => {
```
Comment (JGiter) [inline test/initializeConnection.testSuite.ts:9]: Why do you think this test is not needed anymore?

File: src/iam/iam-base.ts (hunk 6 above)
Comment (dawidgil) [reply src/iam/iam-base.ts]: Ah, checking something and I committed this accidentally. Reverted.

File: test/initializeConnection.testSuite.ts (hunk 7 above)
Comment (dawidgil) [reply test/initializeConnection.testSuite.ts:9]: Because in node env in constructor we are setting providerType always. So it is not possible to have not defined walletProvider.
I was trying to use spyOn this property, but without success.

Comment (JGiter) [reply test/initializeConnection.testSuite.ts:9]: But the test anyway should confirm that error will be thrown when private key is not provided, just with different error message

Comment (dawidgil) [reply test/initializeConnection.testSuite.ts:9]: Ok, i changed error message there.

**Review by JGiter — APPROVED** [review]
